br/pkg/lightning/errormanager/errormanager.go

```diff
@@ -642,11 +651,116 @@ func (em *ErrorManager) ReplaceConflictKeys(
 				}
 			}
 		}
-		if err := rawKeyRows.Err(); err != nil {
+		if err := indexKvRows.Err(); err != nil {
+			return errors.Trace(err)
+		}
+
+		// check data KV
+		dataKvRows, err := em.db.QueryContext(
+			gCtx, fmt.Sprintf(selectDataConflictKeysReplace, em.schemaEscaped),
+			tableName)
+		if err != nil {
 			return errors.Trace(err)
 		}
+		defer dataKvRows.Close()
+
+		var previousRawKey, latestValue []byte
+		var mustKeepKvPairs *kv.Pairs
+
+		for dataKvRows.Next() {
+			var rawKey, rawValue, rawHandle []byte
+			if err := dataKvRows.Scan(&rawKey, &rawValue, &rawHandle); err != nil {
+				return errors.Trace(err)
+			}
+			em.logger.Debug("got group raw_key, raw_value, raw_handle from table",
+				logutil.Key("raw_key", rawKey),
+				zap.Binary("raw_value", rawValue),
+				zap.Binary("raw_handle", rawHandle))
 
-		// TODO: check data KV
+			if !bytes.Equal(rawKey, previousRawKey) {
+				previousRawKey = rawKey
+				// get the latest value of rawKey from downstream TiDB
+				latestValue, err = fnGetLatest(gCtx, rawKey)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if latestValue != nil {
+					handle, err := tablecodec.DecodeRowKey(rawKey)
+					if err != nil {
+						return errors.Trace(err)
+					}
+					decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
+						tbl.Meta(), handle, tbl.Cols(), latestValue)
+					if err != nil {
+						return errors.Trace(err)
+					}
+					_, err = encoder.Table.AddRecord(encoder.SessionCtx, decodedData)
+					if err != nil {
+						return errors.Trace(err)
+					}
+					// calculate the new mustKeepKvPairs corresponding to the new rawKey
+					// find out all the KV pairs that are contained in the data KV
+					mustKeepKvPairs = encoder.SessionCtx.TakeKvPairs()
+				}
+			}
+
+			// if the latest value of rawKey equals to rawValue, that means this data KV is maintained in downstream TiDB
+			// if not, that means this data KV has been deleted due to overwritten index KV
+			if bytes.Equal(rawValue, latestValue) {
+				continue
+			}
+
+			handle, err := tablecodec.DecodeRowKey(rawHandle)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
+				tbl.Meta(), handle, tbl.Cols(), rawValue)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			_, err = encoder.Table.AddRecord(encoder.SessionCtx, decodedData)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			// find out all the KV pairs that are contained in the data KV
+			kvPairs := encoder.SessionCtx.TakeKvPairs()
+			for _, kvPair := range kvPairs.Pairs {
+				em.logger.Debug("got encoded KV",
+					logutil.Key("key", kvPair.Key),
+					zap.Binary("value", kvPair.Val))
+				kvLatestValue, err := fnGetLatest(gCtx, kvPair.Key)
+				if tikverr.IsErrNotFound(err) || kvLatestValue == nil {
```

Comment:
seems `fnGetLatest` will return error which will be checked by `IsErrNotFound`. For above usage of `fnGetLatest` do we need to ignore the not-found error and continue the function?

Reply:
according to the demo code, here if the value of `kvPair.Key` is not found, that means `kvPair.Key` has been deleted, so we need to continue in the for loop
https://github.com/lyzx2001/tidb-conflict-replace/blob/9666d86f660f9a91e4a0446027bf0319a88abf2a/src/main.go#L178-L180

Reply:
Why do we return for not-found error in line 863

Reply:
thanks for reminding, we should first check whether `err` is not-found error